lib/dynflow/world.rb

```diff
@@ -1,41 +1,50 @@
+# -*- coding: utf-8 -*-
 module Dynflow
   class World
     include Algebrick::TypeCheck
+    include Algebrick::Matching
 
-    attr_reader :executor, :persistence, :transaction_adapter, :action_classes, :subscription_index,
-                :logger_adapter, :options, :middleware, :auto_rescue
-
-    def initialize(options_hash = {})
-      @options             = default_options.merge options_hash
-      @logger_adapter      = Type! option_val(:logger_adapter), LoggerAdapters::Abstract
-      @transaction_adapter = Type! option_val(:transaction_adapter), TransactionAdapters::Abstract
-      persistence_adapter  = Type! option_val(:persistence_adapter), PersistenceAdapters::Abstract
-      @persistence         = Persistence.new(self, persistence_adapter)
-      @executor            = Type! option_val(:executor), Executors::Abstract
-      @action_classes      = option_val(:action_classes)
-      @auto_rescue         = option_val(:auto_rescue)
-      @exit_on_terminate   = option_val(:exit_on_terminate)
-      @middleware          = Middleware::World.new
+    attr_reader :id, :client_dispatcher, :executor_dispatcher, :executor, :connector,
+        :transaction_adapter, :logger_adapter, :coordinator,
+        :persistence, :action_classes, :subscription_index,
+        :middleware, :auto_rescue, :clock
+
+    def initialize(config)
+      @id                   = SecureRandom.uuid
+      @clock                = Clock.spawn('clock')
+      config_for_world      = Config::ForWorld.new(config, self)
+      config_for_world.validate
+      @logger_adapter       = config_for_world.logger_adapter
+      @transaction_adapter  = config_for_world.transaction_adapter
+      @persistence          = Persistence.new(self, config_for_world.persistence_adapter)
+      @coordinator          = Coordinator.new(config_for_world.coordinator_adapter)
+      @executor             = config_for_world.executor
+      @action_classes       = config_for_world.action_classes
+      @auto_rescue          = config_for_world.auto_rescue
+      @exit_on_terminate    = config_for_world.exit_on_terminate
+      @connector            = config_for_world.connector
```

Comment:
I still tend to link this, where we have all the references needed in one place, some of them we would still need, as they are not part of the configuration. +1 for keeping them